cellrank/tools/_utils.py

```diff
@@ -1150,3 +1150,156 @@ def _info_if_obs_keys_categorical_present(
+        raise ValueError("Rows in `a_fuzzy` do not sum to one.")
+
+    # initialise
+    n_clusters = a_fuzzy.shape[1]
+    if raise_threshold is not None:
+        n_raise = int(raise_threshold * n_most_likely)
+
+    # create argmax assignment as a lookup table
+    max_assignment = a_fuzzy.argmax(axis=1)
+
+    # initially select `n_most_likely` samples per cluster
+    sample_assignment = {
+        cl: fuzzy_assignment.argsort()[::-1][:n_most_likely]
+        for cl, fuzzy_assignment in enumerate(a_fuzzy.T)
+    }
+
+    # create the one-hot encoded discrete clustering
+    a_discrete = np.zeros_like(a_fuzzy)
+    for ix in range(n_clusters):
+        a_discrete[sample_assignment[ix], ix] = 1
+
+    # handle samples assigned to more than one cluster
+    critical_samples = np.where(a_discrete.sum(1) > 1)[0]
+    for sample_ix in critical_samples:
+        if remove_overlap:
+            a_discrete[sample_ix, :] = _one_hot(n_clusters)
+        else:
+            a_discrete[sample_ix, :] = _one_hot(n_clusters, max_assignment[sample_ix])
+
+    # check how many samples this left for each cluster
+    n_samples_per_cluster = a_discrete.sum(0)
+    if raise_threshold is not None:
+        if (n_samples_per_cluster < n_raise).any():
+            raise ValueError(
+                f"Discretizing leads to a cluster with less than {n_raise} samples. "
+                f"Consider recomputing the fuzzy clustering."
+            )
+    critical_clusters = np.where(n_samples_per_cluster < n_most_likely)[0]
+
+    return a_discrete, critical_clusters
+
+
+def _series_from_one_hot_matrix(
+    a: np.array, index: Optional[Iterable] = None, names: Optional[Iterable] = None
+) -> pd.Series:
+    """
+    Create a pandas Series based on a one-hot encoded matrix.
+
+    Params
+    ------
+    a
+        One-hot encoded membership matrix, of shape (`n_samples x n_clusters`) i.e. a `1` in position `i, j`
+        signifies that sample `i` belongs to cluster `j`.
+    index
+        Index for the Series. Careful, if this is not given, categories are removed when writing to AnnData.
+
+    Returns
+    -------
+    cluster_series
+        Pandas Series, indicating cluster membership for each sample. The dtype of the categories is `str`
+        and samples that belong to no cluster are assigned `NaN`.
+    """
+    n_clusters, n_samples = a.shape[1], a.shape[0]
+    if not type(a) == np.ndarray:
+        raise TypeError(
+            f"Expected `a` to be of type `numpy.ndarray`, got `{type(a).__name__!r}`."
+        )
+    if index is None:
+        index = range(n_samples)
+    if names is not None:
+        if len(names) != n_clusters:
+            raise ValueError(
+                f"Shape mismatch, length of `names` is `{len(names)}`, but `n_clusters` = {n_clusters}."
+            )
+    else:
+        names = np.arange(n_clusters).astype("str")
+
+    target_series = pd.Series(index=index, dtype="category")
```

Comment:
How efficient is this?

Reply:
I don't think it's really going to matter since the number of states is always very low.